### ISSUE TITLE
ERROR: builtin_op_importers.cpp:2179 In function importPad: [8] Assertion failed: inputs.at(1).is_weights()

### ISSUE THREAD

**Issue by HelloAlone** [NVIDIA/TensorRT#670]
## Description

My current workflow is pb -> onnx -> tensorrt. Thanks to @jignparm , (refer [https://github.com/onnx/tensorflow-onnx/issues/994](https://github.com/onnx/tensorflow-onnx/issues/994)) I finally converted the original pb to onnx model. But I received a importPad error when converting onnx to tensorrt engine. It seems is a TensorRT issue according to jignparm.

> [V] [TRT] ModelImporter.cpp:125: preprocessor/PadV2 [Pad] inputs: [preprocessor/resize/ResizeBilinear:0 -> (-1, -1, -1, -1)], [preprocessor/PadV2__135:0 -> (8)], [preprocessor/PadV2/constant_values:0 -> ()],
ERROR: builtin_op_importers.cpp:2179 In function importPad:
[8] Assertion failed: inputs.at(1).is_weights()

## Environment

**TensorRT Version**: 7.1.3.4 GA
**GPU Type**: GTX 1080Ti
**Nvidia Driver Version**: 440.33.01
**CUDA Version**: 10.2
**CUDNN Version**: 8
**Operating System + Version**: Ubuntu 16.04
**Python Version (if applicable)**: 3.7
**TensorFlow Version (if applicable)**: 1.15
**Onnx Version (if applicable)**: 1.6.0
**TensorFlow-onnx Version (if applicable)**: 1.6.2

## Relevant Files

The onnx file can be download at [https://yadi.sk/d/NW5no-n0HtNECw](https://yadi.sk/d/NW5no-n0HtNECw). The original pb file can be download at [https://yadi.sk/d/PN3tDlm6GzwSpw](https://yadi.sk/d/PN3tDlm6GzwSpw)

## Steps To Reproduce

1. edit the pb and set the initializer to a type TRT can handle (float64->float32):

```python
import tensorflow as tf
from tensorflow.python.compiler.tensorrt import trt_convert as trt
import numpy as np
from PIL import Image
with tf.Session() as sess:
    # First deserialize your frozen graph:
    with tf.gfile.GFile('models/rjcd_13_0.1421_0.1290.pb', 'rb') as f:
        frozen_graph = tf.GraphDef()
        frozen_graph.ParseFromString(f.read())
    with tf.Graph().as_default() as graph:
        tf.import_graph_def(frozen_graph, name='')
    new_model = tf.GraphDef()
    with tf.Session(graph=graph) as sess:
        for n in sess.graph_def.node:
            if n.name == 'resize_information_computer/truediv/Cast' or n.name == 'resize_information_computer/truediv/Cast_1':
                nn = new_model.node.add()
                nn.CopyFrom(n)
                nn.attr['DstT'].CopyFrom(tf.AttrValue(type=tf.float32.as_datatype_enum))
            elif n.name == 'resize_information_computer/truediv':
                nn = new_model.node.add()
                nn.CopyFrom(n)
                nn.attr['T'].CopyFrom(tf.AttrValue(type=tf.float32.as_datatype_enum))
            elif n.name == 'resize_information_computer/ToFloat':
                nn = new_model.node.add()
                nn.CopyFrom(n)
                nn.attr['SrcT'].CopyFrom(tf.AttrValue(type=tf.float32.as_datatype_enum))
            else:
                nn = new_model.node.add()
                nn.CopyFrom(n)
    with tf.gfile.GFile('models/rjcd_13_0.1421_0.1290_new.pb', mode='wb') as f:
        f.write(new_model.SerializeToString())
```

2. convert pb to onnx

```
python -m tf2onnx.convert --graphdef rjcd_13_0.1421_0.1290_new.pb --output rjcd_13_0.1421_0.1290.onnx --inputs image_input:0,max_detections:0,iou_threshold:0,score_threshold:0 --outputs output/Squeeze:0 --opset 11 --fold_const
```

3. convert onnx to tensorrt engine

```
trtexec --onnx=rjcd_13_0.1421_0.1290.onnx --explicitBatch
```

**Comment by XinnWang** [comment]
i met the same problem. have you solved it ?

**Comment by zccyman** [comment]
have same problem.

**Comment by lrh454830526** [comment]
Any idea to solve?

**Comment by lrh454830526** [comment]
Have you solved the problem?

**Comment by hewumars** [comment]
I met the same problem.have you solved it ?
ERROR: builtin_op_importers.cpp:436 In function importConv: Assertion failed: inputs.at(2).is_weights()

**Comment by XinnWang** [comment]
https://github.com/TRI-ML/packnet-sfm/issues/52


看看这个例子对你有没有帮助吧。




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "NVIDIA/TensorRT"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年8月26日(星期三) 下午5:01
收件人:&nbsp;"NVIDIA/TensorRT"<TensorRT@noreply.github.com&gt;;
抄送:&nbsp;"潇"<616914658@qq.com&gt;;"Comment"<comment@noreply.github.com&gt;;
主题:&nbsp;Re: [NVIDIA/TensorRT] ERROR: builtin_op_importers.cpp:2179 In function importPad: [8] Assertion failed: inputs.at(1).is_weights() (#670)





 
I met the same problem.have you solved it ?
 ERROR: builtin_op_importers.cpp:436 In function importConv: Assertion failed: inputs.at(2).is_weights()

**Comment by hyc-xyz** [comment]
same problem

**Comment by cognitiveRobot** [comment]
Getting the same error. Any update?
```
ERROR: builtin_op_importers.cpp:2179 In function importPad:
[8] Assertion failed: inputs.at(1).is_weights()
[09/28/2020-13:43:43] [E] Failed to parse onnx file
[09/28/2020-13:43:43] [E] Parsing model failed
[09/28/2020-13:43:43] [E] Engine creation failed
[09/28/2020-13:43:43] [E] Engine set up failed
```

**Comment by mk-nvidia** [comment]
Thanks for reporting, we'll take a look.

**Comment by simutisernestas** [comment]
facing same issue

**Comment by SteveMacenski** [comment]
Anyone find a solution to this?

**Comment by batrlatom** [comment]
internet if flooding with the requests for solution :] . this would help to convert maskrnn, fasterrcnn, retinanet and many other networks.

**Comment by TomWildenhain-Microsoft** [comment]
The assertion suggests that TensorRT requires the pads to be constant (known at model load time).  In the ONNX model, this is not the case:

![image](https://user-images.githubusercontent.com/67606533/102656827-8631f500-4142-11eb-96a3-3da4735f4784.png)

The pads are computed by this logic:

![image](https://user-images.githubusercontent.com/67606533/102656864-9a75f200-4142-11eb-8081-8eb88517f493.png)

This is partially be because the input size itself is unknown:
![image](https://user-images.githubusercontent.com/67606533/102656977-d14c0800-4142-11eb-9dfb-264235f9e1ec.png)

There are 3 possible ways to get this model to work:
1.  Update TensorRT to support non-const pads (as per ONNX spec)
2. Update tf2onnx to convert non-const pads to a composition of other ops with the same behavior (Const of shape, concat) IF TensorRT is the conversion target.  May cause perf issues
3. Choose fixed image dims for the input and reconvert the model using those dims.  I haven't tried it, but the calculations should fold away into constants.

**Comment by adjagetia** [comment]
Solved it using [onnx-simplifier](https://github.com/daquexian/onnx-simplifier)

**Comment by ilya-SX** [comment]
> Solved it using [onnx-simplifier](https://github.com/daquexian/onnx-simplifier)

Hi. Can you please share which network did you use and how did you solve the issue?
I am using faster rcnn and can't find a solution to it.

**Comment by O2andsAnoxia** [comment]
> > Solved it using [onnx-simplifier](https://github.com/daquexian/onnx-simplifier)
> 
> Hi. Can you please share which network did you use and how did you solve the issue?
> I am using faster rcnn and can't find a solution to it.

@ilya-SX  Did you implemented the conversion of faster-rcnn? Could you tell me how to do ,if your problem was solved.thanks!

**Comment by O2andsAnoxia** [comment]
@ilya-SX  Did you implemented the conversion of faster-rcnn? Could you tell me how to do ,if your problem was solved.I solves this problem using onnx-simplifier,but there are a new problem as follow:

[04/21/2021-15:11:37] [W] [TRT] onnx2trt_utils.cpp:220: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
[04/21/2021-15:11:37] [W] [TRT] onnx2trt_utils.cpp:246: One or more weights outside the range of INT32 was clamped
[04/21/2021-15:11:37] [E] [TRT] **onnx2trt_utils.cpp:467: Found unsupported datatype (11) when importing initializer: 2739
ERROR: ModelImporter.cpp:92 In function parseGraph:
[8] Assertion failed: convertOnnxWeights(initializer, &weights, ctx)**
[04/21/2021-15:11:37] [E] Failed to parse onnx file
[04/21/2021-15:11:37] [E] Parsing model failed
[04/21/2021-15:11:37] [E] Engine creation failed
[04/21/2021-15:11:37] [E] Engine set up failed

**Comment by ilya-SX** [comment]
@O2andsAnoxia Unfortunately I didn't find a solution to the problem. My results were similar to yours.

**Comment by O2andsAnoxia** [comment]
> @O2andsAnoxia Unfortunately I didn't find a solution to the problem. My results were similar to yours.

what a pity! Maybe we can discuss these problem,if you don't mind.This is my QQ(1766529899)

**Comment by wkl2013DeepVision** [comment]
same too!

ONNX IR version:  0.0.6
Opset version:    11
Producer name:    pytorch
Producer version: 1.7
Domain:
Model version:    0
Doc string:
----------------------------------------------------------------
[05/14/2021-08:44:03] [W] [TRT] onnx2trt_utils.cpp:220: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
[05/14/2021-08:44:03] [W] [TRT] onnx2trt_utils.cpp:246: One or more weights outside the range of INT32 was clamped
ERROR: builtin_op_importers.cpp:2249 In function importPad:
[8] Assertion failed: inputs.at(1).is_weights()
Trying to use DLA core 0 on a platform that doesn't have any DLA cores
[05/14/2021-08:44:03] [E] [TRT] Network must have at least one output
[05/14/2021-08:44:03] [E] [TRT] Network validation failed.

**Comment by ttyio** [comment]
@HelloAlone  Have you tried 

    polygraphy surgeon sanitize --fold-constants

see https://github.com/onnx/onnx-tensorrt/blob/master/docs/faq.md#common-assertion-errors, thanks

**Comment by ttyio** [comment]
Closing since no activities for more than 3 weeks, please reopen if you still have question. Thanks!

**Comment by phamdat09** [comment]
@ttyio I already tried with `polygraphy surgeon sanitize --fold-constants`. But it is still raise same error: `In function importPad:
[8] Assertion failed: inputs.at(1).is_weights() && "The input pads is required to be an initializer."`

**Comment by liuanhua110** [comment]
Why this issuse is closed ? STILL WRONG

**Comment by jucic** [comment]
Why this issuse is closed ? STILL WRONG!!!

**Comment by TomWildenhain-Microsoft** [comment]
> @ttyio I already tried with `polygraphy surgeon sanitize --fold-constants`. But it is still raise same error: `In function importPad: [8] Assertion failed: inputs.at(1).is_weights() && "The input pads is required to be an initializer."`

Looks like polygraphy also requires constant pads. Try inspecting the model to see if the pads will become constant when the input shape is constant. If so, assign the input shape and try onnx simplifier.

**Comment by ttyio** [comment]
Hello @TomWildenhain-Microsoft , could you try TRT 8.2? we now support non constant padding, thanks!

**Comment by ttyio** [comment]
@phamdat09  @liuanhua110 @jucic Have you tried 8.2 TRT? we support non constant padding now, thanks!

**Comment by wkl2013DeepVision** [comment]
yes , 8.2TRT works well! thank you~



At 2022-01-13 10:23:58, "Xiaodong (Vincent) Huang" ***@***.***> wrote:

@***@***.***@jucic Have you tried 8.2 TRT? we support non constant padding now, thanks!

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by wkl2013DeepVision** [comment]
yes , 8.2TRT works well! thank you~



At 2022-01-12 09:49:32, "jucic" ***@***.***> wrote:

Why this issuse is closed ? STILL WRONG!!!

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by wkl2013DeepVision** [comment]
yes , 8.2TRT works well! thank you~



At 2022-01-13 10:21:44, "Xiaodong (Vincent) Huang" ***@***.***> wrote:

Hello @TomWildenhain-Microsoft , could you try TRT 8.2? we now support non constant padding, thanks!

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by wkl2013DeepVision** [comment]
yes , 8.2TRT works well! thank you~



At 2022-01-13 02:06:54, "TomWildenhain-Microsoft" ***@***.***> wrote:

@ttyio I already tried with polygraphy surgeon sanitize --fold-constants. But it is still raise same error: In function importPad: [8] Assertion failed: inputs.at(1).is_weights() && "The input pads is required to be an initializer."

Looks like polygraphy also requires constant pads. Try inspecting the model to see if the pads will become constant when the input shape is constant. If so, assign the input shape and try onnx simplifier.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by duanshengliu** [comment]
> Solved it using [onnx-simplifier](https://github.com/daquexian/onnx-simplifier)

Thanks, my problem has solved!

**Comment by duanshengliu** [comment]
using onnx-simplifier solved my problem, thanks

**Comment by wkl2013DeepVision** [comment]
thank you



在 2022-06-13 20:16:35，"duanshengliu" ***@***.***> 写道：

using onnx-simplifier solved my problem, thanks
